packages/lib/src/project/AppProject.js

```diff
@@ -1,26 +1,36 @@
 import BasePackageProject from "./BasePackageProject";
 import App from "../app/App";
 import Package from "../package/Package";
+import { DeployError } from '../utils/errors/DeployError';
 import _ from 'lodash';
 
 export default class AppProject extends BasePackageProject {
-  static async fetch(appAddress, name, txParams) {
-    const app = await App.fetch(appAddress, txParams)
-    const packageInfo = await app.getPackage(name)
-    const project = new this(app, name, packageInfo.version, txParams)
-    project.package = packageInfo.package
-    return project
-  }
-
-  static async deploy(name = 'main', version = '0.1.0', txParams = {}) {
-    const thepackage = await Package.deploy(txParams)
-    const directory = await thepackage.newVersion(version)
-    const app = await App.deploy(txParams)
-    await app.setPackage(name, thepackage.address, version)
-    const project = new this(app, name, version, txParams)
-    project.directory = directory
-    project.package = thepackage
-    return project
+  static async fetchOrDeploy(name = 'main', version = '0.1.0', txParams = {}, { appAddress = undefined, packageAddress = undefined }) {
+    let thepackage, directory, app
+    try {
+      app = appAddress
+        ? await App.fetch(appAddress, txParams)
+        : await App.deploy(txParams)
+      if (packageAddress) {
+        thepackage = await Package.fetch(packageAddress, txParams)
+      } else if (await app.hasPackage(name, version)) {
+        thepackage = (await app.getPackage(name)).package
+      } else {
+        thepackage = await Package.deploy(txParams)
+      }
+      directory = await thepackage.hasVersion(version)
+        ? await thepackage.getDirectory(version)
+        : await thepackage.newVersion(version)
+
+      if (!await app.hasPackage(name, version)) await app.setPackage(name, thepackage.address, version)
+      const project = new this(app, name, version, txParams)
+      project.directory = directory
+      project.package = thepackage
+
+      return project
+    } catch(deployError) {
+      throw new DeployError(deployError.message, { thepackage, directory, app })
+    }
```

Comment:
What do you think about splitting this method a bit to sth like:
```js
  static async fetchOrDeploy(name = 'main', version = '0.1.0', txParams = {}, { appAddress = undefined, packageAddress = undefined }) {
    try {
      const app = await this.fetchOrDeployApp(...)
      const thepackage = await this.fetchOrDeployPackage(...)
      const directory = await this.fetchOrDeployDirectory(...)
      return this.buildProject(...)
    } catch(deployError) {
      throw new DeployError(deployError.message, { thepackage, directory, app })
    }
  }
```

Reply:
We've discussed this offline. It's not a good idea to merge `deploy` and `fetch` methods in `Package` and `App`, as it would add complexity. 
On the other hand, separating the logic to new static methods inside `AppProject`/`LibProject` (or `BasePackageProject`) just to make the code prettier is not reason enough to do so. In case we would want to provide this methods through `zos-lib`, we can add them, but this exceeds this PR's scope.